### PR TITLE
Add Communicator::waitForShutdownAsync in C++

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR adds Communicator::waitForShutdownAsync in C++ and uses this function to implement Communicator.shutdownCompleted() in Swift.

### CODE REVIEW

**Review by pepone — APPROVED** [review]

File: cpp/src/Ice/ObjectAdapterFactory.cpp [hunk 1]
```diff
@@ -39,59 +37,116 @@ IceInternal::ObjectAdapterFactory::shutdown()
...
+                                                            callback();
+                                                        }
+                                                    }};
+
+                // Joining this thread is challenging because a callback can destroy the communicator and indirectly
+                // this object adapter factory. So we just detach it.
+                shutdownCompletedThread.detach();
+            }
+
+            _shutdownCompletedCallbacks.emplace_back(std::move(completed));
```
Comment (externl) [inline cpp/src/Ice/ObjectAdapterFactory.cpp:111]: I think std::push_back would be more idiomatic here since we're moving an existing function, not creating a new one.

**Review by externl — APPROVED** [review]

File: cpp/src/Ice/ObjectAdapterFactory.cpp (hunk 1 above)
Comment (bernardnormier) [reply cpp/src/Ice/ObjectAdapterFactory.cpp]: Fixed

File: cpp/src/Ice/ObjectAdapterFactory.cpp [hunk 2]
```diff
@@ -39,59 +37,128 @@ IceInternal::ObjectAdapterFactory::shutdown()
...
+                                                           self->waitForShutdown();
+                                                           vector<function<void()>> callbacks;
+                                                           {
+                                                               lock_guard lg(self->_mutex);
+                                                               callbacks = std::move(self->_shutdownCompletedCallbacks);
+                                                           }
+
+                                                           for (const auto& callback : callbacks)
+                                                           {
+                                                               callback();
```
Comment (externl) [inline cpp/src/Ice/ObjectAdapterFactory.cpp:103]: We should probably try/catch around these user callbacks.

File: cpp/src/Ice/ObjectAdapterFactory.cpp [hunk 3]
```diff
@@ -39,59 +37,128 @@ IceInternal::ObjectAdapterFactory::shutdown()
...
+    if (shutdownCompletedThread.joinable())
+    {
+        if (shutdownCompletedThread.get_id() == std::this_thread::get_id())
+        {
+            // destroy is called from a shutdown completed callback - this is fine, but the thread can't join itself.
+            shutdownCompletedThread.detach();
+        }
+        else
+        {
+            shutdownCompletedThread.join();
```
Comment (externl) [inline cpp/src/Ice/ObjectAdapterFactory.cpp:152]: Could we just join the tread in the ObjectAdapterFactory destructor?

**Review by externl — APPROVED** [review]

File: cpp/src/Ice/ObjectAdapterFactory.cpp (hunk 2 above)
Comment (bernardnormier) [reply cpp/src/Ice/ObjectAdapterFactory.cpp:103]: The doc-comment says the callback must not throw any exception.

File: cpp/src/Ice/ObjectAdapterFactory.cpp (hunk 3 above)
Comment (bernardnormier) [reply cpp/src/Ice/ObjectAdapterFactory.cpp:152]: It's nicer to join it in destroy (= earlier), and we would have the same join/detach issue in the destructor: the last refcount can be released by the shutdown completed thread.

**Review by pepone — APPROVED** [review]
